src/framework/components/element/markup.js

```diff
@@ -0,0 +1,417 @@
+                    this._error = "expected string";
+                    return false;
+                }
+                tag.value = this._scanner.buf().join("");
+                token = this._scanner.read();
+            }
+
+            // read optional tag attributes
+            while (true) {
+                switch (token) {
+                    case this._scanner.CLOSE_BRACKET_TOKEN:
+                        tags.push(tag);
+                        return true;
+                    case this._scanner.IDENTIFIER_TOKEN:
+                        var identifier = this._scanner.buf().join("");
+                        token = this._scanner.read();
+                        if (token !== this._scanner.EQUALS_TOKEN) {
+                            this._error = "expected equals";
+                            return false;
+                        }
+                        token = this._scanner.read();
+                        if (token !== this._scanner.STRING_TOKEN) {
+                            this._error = "expected string";
+                            return false;
+                        }
+                        var value = this._scanner.buf().join("");
+                        tag.attributes[identifier] = value;
+                        break;
+                    default:
+                        this._error = "expected close bracket or identifier";
+                        return false;
+                }
+                token = this._scanner.read();
+            }
+        }
+    });
+
+    // copy the contents of source into target and map the values with the
+    // provided mapping table.
+    function mapAndMerge(target, source) {
+        for (var key in source) {
+            if (!source.hasOwnProperty(key)) {
+                continue;
+            }
+            var value = source[key];
+            if (value instanceof Object) {
+                if (!target.hasOwnProperty(key)) {
+                    target[key] = { };
+                }
+                mapAndMerge(target[key], source[key]);
+            } else {
+                target[key] = value;
+            }
+        }
+    }
+
+    // get the markup tags which apply to the given symbol. all tags overlapping the
+    // symbol must be composed into a single structure.
+    function resolveMarkupTags(tags, symbolIndex) {
+        var result = { };
+        for (var index = 0; index < tags.length; ++index) {
+            var tag = tags[index];
+            if (tag.start <= symbolIndex &&
+                (tag.end === null || tag.end > symbolIndex)) {
+                var tmp = { };
+                tmp[tag.name] = { value: tag.value, attributes: tag.attributes };
+                mapAndMerge(result, tmp);
+            }
+        }
+        return result;
+    }
+
+    // evaluate the list of symbols, extract the markup tags and return an
+    // array of symbol and tag pairs
+    function evaluateMarkup(symbols) {
```

Comment:
My main worry about this function is that it will allocate a lot of objects and since it's called on every updateText, it might end up being expensive if used for text that is updated every frame e.g. timers / scores etc. It might be useful to maybe have a single parser / scanner as well? I realize this might not be possible...

Reply:
I think the scanning and parsing is about as optimal as it can be (apart from possible micro-optimizations). However the per-character tag assignment creates more objects than necessary, will try improve this.